client/src/components/related/Outfit.jsx

```diff
@@ -2,57 +2,95 @@ import React from 'react';
 import ProductCard from './ProductCard.jsx';
 import axios from 'axios';
 
-const Outfit = (props) => {
-  const { selectedProduct, selectProduct, outfit, updateOutfit } = props;
-  const { localStorage } = window;
+class Outfit extends React.Component {
+  constructor(props) {
+    super(props);
 
-  const addToOutfit = () => {
-    if (outfit && outfit.includes(selectedProduct)) {
-      return;
+    this.addToOutfit = this.addToOutfit.bind(this);
+    this.removeFromOutfit = this.removeFromOutfit.bind(this);
+
+    this.state = {
+      outfit: []
+    };
+  }
+
+  loadOutfit() {
+    const { localStorage } = window;
+    const outfitData = localStorage.getItem('outfit');
+
+    if (outfitData) {
+      const outfit = JSON.parse(outfitData);
+      console.log('Outfit found in localStorage:', outfit);
+      this.setState({ outfit: outfit });
+    }
+  }
+
+  addToOutfit() {
+    const { localStorage } = window;
+    const { selectedProduct } = this.props;
+    const { outfit } = this.state;
+
+    if (!outfit.includes(selectedProduct.id)) {
+      console.log(`Add ${selectedProduct.id} (${selectedProduct.name}) to outfit`);
+      let newOutfit = [ selectedProduct.id, ...outfit ];
+      localStorage.setItem('outfit', JSON.stringify(newOutfit));
```

Comment:
Just a general observation/question of what is the most optimal: You are using caching/local storage to reduce load from API calls, but I'm sure JSON.parse & JSON.stringify are expensive operations as well! Certainly much more so than switching references in memory. I see enough of these calls going on that I wonder if this method is only faster in cases of smaller data sets, and if other methods or living with less caching is more optimal in other situations, and what the dividing line is between these?

Reply:
`window.localStorage` allows any data type you want as long as it's a string.